src/espidf.rs

```diff
@@ -105,47 +94,87 @@ impl Tools {
     }
 }
 
-/// Installer for the esp-idf source and tools.
+/// Information about a esp-idf source and tools installation.
+pub struct EspIdf {
+    /// The esp-idf repository.
+    pub esp_idf: git::Repository,
+    /// The binary paths of all tools concatenated with the system `PATH` env variable.
+    pub exported_path: OsString,
+}
+
+impl EspIdf {
+    pub fn detect_from_path() -> Result<Option<Self>> {
+        if cmd!["python", "idf.py", "--help"].is_ok() {
+            let version =
+                EspIdfVersion::try_from_env_var()?.ok_or_else(|| anyhow::anyhow!("TODO"))?;
+
+            let idf = Self {
+                esp_idf: version,
+                exported_path: env::var_os("PATH").unwrap_or_else(OsString::new),
+            };
+
+            Ok(Some(idf))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+/// EspIdfVersion can be either a managed or unmanaged one.
+///
+/// The managed one is represented by a GIT url and a git Ref. The URL is optional, and if not provided
+/// will default to the main ESP-IDF GitHub repository.
+///
+/// An unmanaged ESP-IDF version is represented by a user-provided local clone of ESP-IDF.
+///
+/// The main difference between managed and unmanaged ESP-IDF versions is reflected in their naming:
+/// - EspIdfVersion::Managed values are cloned locally by the Installer instance, inside its tooling installation directory.
+///   Consenquently, these ESP-IDF repository clones will disappar if the installation directory is deleted by the user.
+/// - EspIdfVersion::Unmanaged values are designating a user-provided, already cloned ESP-IDF repository which lives
+///   outisde the Installer's installation directory. It is only read by the Installer so as the required tooling
+///   for this ESP-IDF repository to be installed.
 #[derive(Debug, Clone)]
-pub struct Installer {
-    version: git::Ref,
-    git_url: Option<String>,
-    local_install_dir: Option<PathBuf>,
-    opts: InstallOpts,
-    tools: Vec<Tools>,
+pub enum EspIdfVersion {
+    Managed((Option<String>, git::Ref)),
+    Unmanaged(git::Repository),
 }
 
-bitflags! {
-    /// Installation options for [`Installer`].
-    pub struct InstallOpts: u32 {
-        const FIND_PREFER_GLOBAL = (1 << 0);
-        const NO_GLOBAL_INSTALL = (1 << 1);
+impl EspIdfVersion {
+    pub fn try_from_env_var() -> Result<Option<git::Repository>> {
+        let version = match env::var(IDF_PATH_VAR) {
+            Err(env::VarError::NotPresent) => None,
+            v => Some(git::Repository::open(v?)?),
+        };
+
+        Ok(version)
+    }
+
+    pub fn from_version_str(
+        url: Option<impl Into<String>>,
+        version: impl AsRef<str>,
+    ) -> (Option<String>, git::Ref) {
+        (
+            url.map(Into::into),
+            decode_esp_idf_version_ref(version.as_ref()),
+        )
     }
```

Comment:
> I think the return values of these functions are pretty awkward, they should really return their own type instead of its parts. To achieve this, I think you should move the logic in `try_from_env_var` into `EspIdf::detect_from_path`. `try_from_env_var` does not really have utility on its own in my opinion.


> Additionally, I think this struct should not be called `EspIdfVersion` (it doesn't contain the version of the esp-idf). The previous names for the `IspIdf::esp_idf_version` and `Installer::version` fields also weren't that great in retrospect. I think a better name would be for example `EspIdfType`.

`EspIdfVersion` is pretty bad indeed, but is `EspIdfType` is better either? This is not a type of an esp-idf repo, but a type of a **reference** to an esp-idf repo. The esp-idf repo is a... repo. That is, ultimately it gets resolved to a `git::Repository`. And moreover, the type notion is encoded in the enum itself (managed vs unmanaged).

So how about:
* EspIdfReference
* EspIdfDescriptor

?

From these two, I have a slight preference for `EspIdfReference`.


> Having a e.g. `esp_idf_type` field with this struct in `EspIdf` would also be good.

I am a bit against exposing all sorts of state for which we do not have a valid use case in the structures. This is why I removed everything which was not used from `EspIdf` and `Pio` (that is, besides the fact that some of the original state was difficult to impossible to construct from an activated ESP-IDF or Pio environment). There is already a resolved reference to an esp-idf repo in `EspIdf`. Why would we need another variation of this information?

Reply:
> I think the return values of these functions are pretty awkward, they should really return their own type instead of its parts. To achieve this, I think you should move the logic in `try_from_env_var` into `EspIdf::detect_from_path`. `try_from_env_var` does not really have utility on its own in my opinion.
> 

It is a public API used in the native builder of `esp-idf-sys` [here](https://github.com/esp-rs/esp-idf-sys/pull/42/files#diff-41d39a20c88287b3858622367bf4dca53863d1145fa87e97f3e4093c386ce1beR359) unfortunately. I can just remove this method and just move its logic in the `esp-idf-sys` crate instead. But then, I should do the same for `from_version_str`. Given that I've accidentally forgotten `decode_esp_idf_version_ref` as public, that would work.

Reply:
Done. `EspIdfVersion` is now `EspIdfReference`. Also, `try_from_env_var` and `from_version_str` are retired.

Reply:
Yup, `EspIdfReference` is a much better name.

> I am a bit against exposing all sorts of state for which we do not have a valid use case in the structures. This is why I removed everything which was not used from `EspIdf` and `Pio` (that is, besides the fact that some of the original state was difficult to impossible to construct from an activated ESP-IDF or Pio environment). There is already a resolved reference to an esp-idf repo in `EspIdf`. Why would we need another variation of this information?

We may not need it (yet), but we may not be the only user of this API. I agree not having to go out of our way to get these, but if we already have them why not give them to the user also? Knowing what reference of the esp-idf it is may be useful to the user.

Reply:
But the user will have the `repository` (= `git::Repository`) field, and the `repository` field is the resolved `esp_idf_reference` anyway. Once the `EspIdfReference` instance is resolved, I don't think the user should use it or reason about it or have access to it. She should be using the `repository` field instead.

Reply:
> But the user will have the `repository` (= `git::Repository`) field, and the `repository` field is the resolved `esp_idf_reference` anyway. Once the `EspIdfReference` instance is resolved, I don't think the user should use it or reason about it or have access to it. She should be using the `repository` field instead.

Agreed for `EspIdfReference`; we can always add it may someone want access to it.